File: qasdk/src/main/java/co/astrnt/qasdk/dao/InformationApiDao.java
package co.astrnt.qasdk.dao;

import org.jetbrains.annotations.NotNull;

import io.realm.RealmList;
import io.realm.RealmObject;
import io.realm.annotations.PrimaryKey;

/**
 * Created by deni rohimat on 19/04/18.
 */
public class InformationApiDao extends RealmObject {

    @PrimaryKey
    private long id;
    private boolean finished;
    private int interviewIndex;
    private int interviewSubIndex;
    private int interviewAttempt;
    private String status;
    private RealmList<PrevQuestionStateApiDao> prevQuestStates;
    private int section_index;
    private int question_index;
    private int preparation_time;
    private int section_duration_left;
    private String section_info;
    private String message;
    private RealmList<QuestionInfoApiDao> questions_info;
    private RealmList<QuestionInfoMcqApiDao> questions_mcq_info;

    public InformationApiDao() {
    }

    public InformationApiDao(int questionIndex, boolean finished, int interviewIndex, int interviewSubIndex, int interviewAttempt, String status, String message, PrevQuestionStateApiDao... prevQuestStates) {
        this.question_index = questionIndex;
        this.finished = finished;
        this.status = status;
        this.interviewIndex = interviewIndex;
        this.interviewSubIndex = interviewSubIndex;
        this.interviewAttempt = interviewAttempt;
        this.message = message;
        if (prevQuestStates != null) {
            this.prevQuestStates = new RealmList<>(prevQuestStates);
        }
    }

    public InformationApiDao(int questionIndex,boolean finished, int interviewIndex, int interviewSubIndex, int interviewAttempt, String status, String message, PrevQuestionStateApiDao[] prevQuestStates, QuestionInfoMcqApiDao... questionInfoMcqApiDao) {
        this.question_index = questionIndex;
        this.finished = finished;
        this.status = status;
        this.interviewIndex = interviewIndex;
        this.interviewSubIndex = interviewSubIndex;
        this.interviewAttempt = interviewAttempt;
        this.message = message;
        if (prevQuestStates != null) {
            this.prevQuestStates = new RealmList<>(prevQuestStates);
        }
        if (questionInfoMcqApiDao != null) {
            this.questions_mcq_info = new RealmList<>(questionInfoMcqApiDao);
        }
    }


    public InformationApiDao(int questionIndex, boolean finished, String status, int section_index, int preparation_time, int section_duration_left, String section_info, String message) {
        this.question_index = questionIndex;
        this.finished = finished;
        this.status = status;
        this.section_index = section_index;
        this.preparation_time = preparation_time;
        this.section_duration_left = section_duration_left;
        this.section_info = section_info;
        this.message = message;
    }

    public InformationApiDao(int questionIndex, boolean finished, String status, int section_index, int preparation_time, int section_duration_left, String section_info, String message, QuestionInfoApiDao... questionInfos) {
        this.question_index = questionIndex;
        this.finished = finished;
        this.status = status;
        this.section_index = section_index;
        this.preparation_time = preparation_time;
        this.section_duration_left = section_duration_left;
        this.section_info = section_info;
        this.message = message;
        if (questionInfos != null) {
            this.questions_info = new RealmList<>(questionInfos);
        }
    }

    public InformationApiDao(int questionIndex, boolean finished, String status, int section_index, int preparation_time, int section_duration_left, String section_info, String message, QuestionInfoMcqApiDao... questionInfoMcqApiDaos) {
        this.question_index = questionIndex;
        this.finished = finished;
        this.status = status;
        this.section_index = section_index;
        this.preparation_time = preparation_time;
        this.section_duration_left = section_duration_left;
        this.section_info = section_info;
        this.message = message;
        if (questionInfoMcqApiDaos != null) {
            this.questions_mcq_info = new RealmList<>(questionInfoMcqApiDaos);
        }
    }

    public long getId() {
        return id;
    }

    public void setId(long id) {
        this.id = id;
    }

    public boolean isFinished() {
        return finished;
    }

    public void setFinished(boolean finished) {
        this.finished = finished;
    }

    public int getInterviewIndex() {
        return interviewIndex;
    }

    public void setInterviewIndex(int interviewIndex) {
        this.interviewIndex = interviewIndex;
    }

    public int getInterviewSubIndex() {
        return interviewSubIndex;
    }

    public void setInterviewSubIndex(int interviewSubIndex) {
        this.interviewSubIndex = interviewSubIndex;
    }

    public int getInterviewAttempt() {
        return interviewAttempt;
    }

    public void setInterviewAttempt(int interviewAttempt) {
        this.interviewAttempt = interviewAttempt;
    }

    public RealmList<PrevQuestionStateApiDao> getPrevQuestStates() {
        return prevQuestStates;
    }

    public void setPrevQuestStates(RealmList<PrevQuestionStateApiDao> prevQuestStates) {
        this.prevQuestStates = prevQuestStates;
    }

    public String getStatus() {
        return status;
    }

    public void setStatus(String status) {
        this.status = status;
    }

    public int getSectionIndex() {
        return section_index;
    }

    public void setSectionIndex(int section_index) {
        this.section_index = section_index;
    }

    public int getPreparationTime() {
        return preparation_time;
    }

    public void setPreparationTime(int preparation_time) {
        this.preparation_time = preparation_time;
    }

    public int getSectionDurationLeft() {
        return section_duration_left;
    }

    public void setSectionDurationLeft(int section_duration_left) {
        this.section_duration_left = section_duration_left;
    }

    public String getSectionInfo() {
        return section_info;
    }

    public void setSectionInfo(String section_info) {
        this.section_info = section_info;
    }

    public String getMessage() {
        return message;
    }

    public void setMessage(String message) {
        this.message = message;
    }

    public RealmList<QuestionInfoApiDao> getQuestionsInfo() {
        return questions_info;
    }

    public void setQuestions_info(RealmList<QuestionInfoApiDao> questions_info) {
        this.questions_info = questions_info;
    }

    public RealmList<QuestionInfoMcqApiDao> getQuestionsMcqInfo() {
        return questions_mcq_info;
    }

    public void setQuestions_mcq_info(RealmList<QuestionInfoMcqApiDao> questions_mcq_info) {
        this.questions_mcq_info = questions_mcq_info;
    }

    public int getQuestion_index() {
        return question_index;
    }

    public void setQuestion_index(int question_index) {
        this.question_index = question_index;
    }

    public boolean isOnGoing() {
        return getSectionInfo().equals("ongoing");
    }

    @NotNull
    @Override
    public String toString() {
        return "InformationApiDao{" +
                "id=" + id +
                ", finished=" + finished +
                ", interviewIndex=" + interviewIndex +
                ", interviewSubIndex=" + interviewSubIndex +
                ", interviewAttempt=" + interviewAttempt +
                ", status='" + status + '\'' +
                ", prevQuestStates=" + prevQuestStates +
                ", section_index=" + section_index +
                ", preparation_time=" + preparation_time +
                ", section_duration_left=" + section_duration_left +
                ", section_info='" + section_info + '\'' +
                ", message='" + message + '\'' +
                ", questions_info=" + questions_info +
                ", questions_mcq_info=" + questions_mcq_info +
                '}';
    }

}

File: qasdk/src/main/java/co/astrnt/qasdk/dao/InformationApiDao.kt
package co.astrnt.qasdk.dao

import io.realm.RealmList
import io.realm.RealmObject
import io.realm.annotations.PrimaryKey

/**
 * Created by deni rohimat on 19/04/18.
 */
open class InformationApiDao : RealmObject {
    @PrimaryKey
    var id: Long = 0
    var isFinished = false
    var interviewIndex = 0
    var interviewSubIndex = 0
    var interviewAttempt = 0
    var status: String? = null
    var prevQuestStates: RealmList<PrevQuestionStateApiDao>? = null
    var sectionIndex = 0
    var question_index = 0
    var preparationTime = 0
    var sectionDurationLeft = 0
    var sectionInfo: String? = null
    var message: String? = null
    var questionsInfo: RealmList<QuestionInfoApiDao>? = null
        private set
    var questionsMcqInfo: RealmList<QuestionInfoMcqApiDao>? = null
        private set

    constructor() {}
    constructor(questionIndex: Int, finished: Boolean, interviewIndex: Int, interviewSubIndex: Int, interviewAttempt: Int, status: String?, message: String?, vararg prevQuestStates: PrevQuestionStateApiDao?) {
        question_index = questionIndex
        isFinished = finished
        this.status = status
        this.interviewIndex = interviewIndex
        this.interviewSubIndex = interviewSubIndex
        this.interviewAttempt = interviewAttempt
        this.message = message
        if (prevQuestStates != null) {
            this.prevQuestStates = RealmList(*prevQuestStates)
        }
    }

    constructor(questionIndex: Int, finished: Boolean, interviewIndex: Int, interviewSubIndex: Int, interviewAttempt: Int, status: String?, message: String?, prevQuestStates: Array<PrevQuestionStateApiDao?>?, vararg questionInfoMcqApiDao: QuestionInfoMcqApiDao?) {
        question_index = questionIndex
        isFinished = finished
        this.status = status
        this.interviewIndex = interviewIndex
        this.interviewSubIndex = interviewSubIndex
        this.interviewAttempt = interviewAttempt
        this.message = message
        if (prevQuestStates != null) {
            this.prevQuestStates = RealmList(*prevQuestStates)
        }
        if (questionInfoMcqApiDao != null) {
            questionsMcqInfo = RealmList(*questionInfoMcqApiDao)
        }
    }

    constructor(interviewIndex: Int, questionIndex: Int, finished: Boolean, status: String?, section_index: Int, preparation_time: Int, section_duration_left: Int, section_info: String?, message: String?) {
        this.interviewIndex = interviewIndex
        question_index = questionIndex
        isFinished = finished
        this.status = status
        sectionIndex = section_index
        preparationTime = preparation_time
        sectionDurationLeft = section_duration_left
        sectionInfo = section_info
        this.message = message
    }

    constructor(questionIndex: Int, finished: Boolean, status: String?, section_index: Int, preparation_time: Int, section_duration_left: Int, section_info: String?, message: String?, vararg questionInfos: QuestionInfoApiDao?) {
        question_index = questionIndex
        isFinished = finished
        this.status = status
        sectionIndex = section_index
        preparationTime = preparation_time
        sectionDurationLeft = section_duration_left
        sectionInfo = section_info
        this.message = message
        if (questionInfos != null) {
            questionsInfo = RealmList(*questionInfos)
        }
    }

    constructor(interviewIndex: Int, interviewSubIndex: Int, questionIndex: Int, finished: Boolean, status: String?, section_index: Int, preparation_time: Int, section_duration_left: Int, section_info: String?, message: String?, vararg questionInfoMcqApiDaos: QuestionInfoMcqApiDao?) {
        this.interviewIndex = interviewIndex
        this.interviewSubIndex = interviewSubIndex
        question_index = questionIndex
        isFinished = finished
        this.status = status
        sectionIndex = section_index
        preparationTime = preparation_time
        sectionDurationLeft = section_duration_left
        sectionInfo = section_info
        this.message = message
        if (questionInfoMcqApiDaos != null) {
            questionsMcqInfo = RealmList(*questionInfoMcqApiDaos)
        }
    }

    fun setQuestions_info(questions_info: RealmList<QuestionInfoApiDao>?) {
        questionsInfo = questions_info
    }

    fun setQuestions_mcq_info(questions_mcq_info: RealmList<QuestionInfoMcqApiDao>?) {
        questionsMcqInfo = questions_mcq_info
    }

    val isOnGoing: Boolean
        get() = sectionInfo == "ongoing"

    override fun toString(): String {
        return "InformationApiDao{" +
                "id=" + id +
                ", finished=" + isFinished +
                ", interviewIndex=" + interviewIndex +
                ", interviewSubIndex=" + interviewSubIndex +
                ", interviewAttempt=" + interviewAttempt +
                ", status='" + status + '\'' +
                ", prevQuestStates=" + prevQuestStates +
                ", section_index=" + sectionIndex +
                ", preparation_time=" + preparationTime +
                ", section_duration_left=" + sectionDurationLeft +
                ", section_info='" + sectionInfo + '\'' +
                ", message='" + message + '\'' +
                ", questions_info=" + questionsInfo +
                ", questions_mcq_info=" + questionsMcqInfo +
                '}'
    }
}